File: backend/src/utils/conversion.py
def excel_col_to_num(col):
    if not col.isalpha(): return False

    return sum((ord(c) - ord('A') + 1) * (26 ** (len(col) - i - 1)) for i, c in enumerate(col))

def num_to_excel_col(num):
    col = ''
    while num > 0:
        num, rem = divmod(num, 26)
        if rem == 0:
            col = 'Z' + col
            num -= 1
        else:
            col = chr(ord('A') + rem - 1) + col
    return col

def coord_to_address(r, c):
    return f'${num_to_excel_col(c+1)}${r+1}'

def address_to_coord(addr):
    cola = ''.join(c for c in addr if c.isalpha())
    row = ''.join(c for c in addr if c.isdigit())
    col = excel_col_to_num(cola)
    row = int(row)
    return row-1, col-1

File: plugin/ohsheet/src/utils/conversion.ts
export function excelColToNum(col: string): number | false {
    if (!/^[A-Za-z]+$/.test(col)) return false;

    return col.split('').reduce((acc, c, i) => {
        return acc + (c.toUpperCase().charCodeAt(0) - 64) * Math.pow(26, col.length - i - 1);
    }, 0);
}

export function numToExcelCol(num: number): string {
    let col = '';
    while (num > 0) {
        let rem = num % 26;
        num = Math.floor(num / 26);
        if (rem === 0) {
            col = 'Z' + col;
            num--;
        } else {
            col = String.fromCharCode(64 + rem) + col;
        }
    }
    return col;
}

export function coordToAddress(r: number, c: number): string {
    return `$${numToExcelCol(c + 1)}$${r + 1}`;
}

export function addressToCoord(addr: string): [number, number] {
    addr = addr.split('!').slice(-1)[0];
    const cola = addr.match(/[A-Za-z]+/)?.[0] || '';
    const row = addr.match(/\d+/)?.[0] || '';
    const col = excelColToNum(cola);
    
    if (col === false) {
        throw new Error('Invalid column reference');
    }
    
    return [parseInt(row) - 1, col - 1];
}
